test-harness

```diff
@@ -3,14 +3,18 @@
 links = './links'
 
 import sys, re, time
+import concurrent.futures, threading, multiprocessing
 from subprocess import Popen, PIPE
 
 successes = failures = ignored = 0
 TIMEOUT = 15 # seconds before a test is timed out.
 
+the_lock = threading.Lock()
```

Comment:
Don't use locks. That is unnecessary complicated. This problem is an **embarrassing parallel problem**. Use a structured programming technique like map-reduce.

Reply:
If I understand Pyhton's GIL correctly, we only need locks for printing, as all changes to shared variables (like successes, failures, ignored) are atomic. 
I'm not sure if a single print("foo") is atomic under the GIL, but it seems so.

We then just need a lock for the three independent print calls in case of a failure such that those three prints don't interleave with another print.
Alternatively, we could just build a string for the overall message, followed by a single print.
This way, we would get rid of all the locking.

Of course, this yields the rather philosophical question of whether we should treat the GIL as an implementation detail or as an intrinsic part of the language. To me it feels wrong to write code with unsynchronized access to shared resources.


In any case, I don't see how using map-reduce would make this particular code less complicated. We would have n reducers running in parallel, working on a list of test cases each.


- If we use idiomatic MR, we would collect the results (e.g., successes, failures, ignored) locally in each reducer rather than globally (as we do now). We would then have to accumulate and return those separate results after all reducers finished.

- I would like to see the results for each test case as soons as they finish, not at the end after all reducers finished. This means that the reducers would have to print the result immediately after each test case. We then face the same problem as with my current solution: Either use a lock to make several print calls atomic or create a single string to be printed and only print once.

- Python doesn't have library support for parallel map-reduce. This means that we would still have to use a pool of threads or processes to parallelize the reducers.

Reply:
It seems there is something like a parallel map, c.f. [https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.multiprocessing.Pool.map](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.multiprocessing.Pool.map).